platform/logger.cpp

```diff
@@ -0,0 +1,273 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(NDEBUG)
+#include <string.h>
+#include <stdlib.h>
+#include "platform/logger_internal.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "platform/mbed_assert.h"
+#if !defined(MBED_DISABLE_ISR_LOGGING) && !defined(MBED_ID_BASED_TRACING)
+#include "platform/CircularBuffer.h"
+#include "rtos/Thread.h"
+
+using namespace rtos;
+using namespace mbed;
+
+// Globals related to ISR logging
+static CircularBuffer <LOG_DATA_TYPE_, LOG_BUF_SIZE_> *log_buffer_;
+#endif
+
+// Globals related to time printing
+static const ticker_data_t *const log_ticker_ = get_us_ticker_data();
+static bool time_enable_ = true;
+
+// Globals related to external buffer capturing
+static LOG_DATA_TYPE_ *extern_buf_ = NULL;
+static uint32_t xbuf_count_ = 0;
+
+static void log_update_buf(const char *data, int32_t size)
+{
+    if ((NULL == extern_buf_) || (NULL == data) || (size <= 0)) {
+        return;
+    }
+    uint32_t count = 0;
+    while(size--)
+    {
+        extern_buf_[xbuf_count_++] = data[count++];
+        xbuf_count_ %= LOG_BUF_SIZE_;
+    }
+}
+
+#if !defined(MBED_DISABLE_ISR_LOGGING) && !defined(MBED_ID_BASED_TRACING)
+extern "C" void log_init()
+{
+    log_buffer_ = new CircularBuffer <LOG_DATA_TYPE_, LOG_BUF_SIZE_>;
+}
+#endif
+
+extern "C" void log_assert(const char *format, ...)
+{
+#if DEVICE_STDIO_MESSAGES
+    core_util_critical_section_enter();
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    va_list args;
+    va_start(args, format);
+    if (time_enable_) {
+        mbed_error_printf("\n[%-8lld]", time);
+    }
+    mbed_error_vfprintf(format, args);
+    va_end(args);
+    mbed_die();
+#endif
+}
+
+extern "C" void log_reset(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = true;
+    xbuf_count_ = 0;
+    core_util_critical_section_exit();
+}
+
+extern "C" void log_buffer_data(LOG_DATA_TYPE_ *str)
+{
+    extern_buf_ = str;
+}
+
+extern "C" void log_disable_time_capture(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = false;
+    core_util_critical_section_exit();
+}
+
+extern "C" void log_enable_time_capture(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = true;
+    core_util_critical_section_exit();
+}
+
+#if defined (MBED_ID_BASED_TRACING)
+
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+extern "C" void print_buf_data(void)
```

Comment:
`#if !defined(MBED_DISABLE_ISR_LOGGING) && !defined(MBED_ID_BASED_TRACING) ` this is the condition
In case of ID based tracing, we use circularBuffer for thread data as well.